node-uniffi/src/types.rs

```diff
@@ -0,0 +1,516 @@
+    /// The current number of established connections.
+    pub num_established: u32,
+    /// The current number of established inbound connections.
+    pub num_established_incoming: u32,
+    /// The current number of established outbound connections.
+    pub num_established_outgoing: u32,
+}
+
+impl From<Libp2pNetworkInfo> for NetworkInfo {
+    fn from(info: Libp2pNetworkInfo) -> Self {
+        Self {
+            num_peers: info.num_peers() as u32,
+            connection_counters: info.connection_counters().into(),
+        }
+    }
+}
+
+impl From<&Libp2pConnectionCounters> for ConnectionCounters {
+    fn from(counters: &Libp2pConnectionCounters) -> Self {
+        Self {
+            num_connections: counters.num_connections(),
+            num_pending: counters.num_pending(),
+            num_pending_incoming: counters.num_pending_incoming(),
+            num_pending_outgoing: counters.num_pending_outgoing(),
+            num_established: counters.num_established(),
+            num_established_incoming: counters.num_established_incoming(),
+            num_established_outgoing: counters.num_established_outgoing(),
+        }
+    }
+}
+
+/// A range of blocks.
+#[derive(Record)]
+pub struct BlockRange {
+    pub start: u64,
+    pub end: u64,
+}
+
+impl From<LuminaBlockRange> for BlockRange {
+    fn from(range: LuminaBlockRange) -> Self {
+        Self {
+            start: *range.start(),
+            end: *range.end(),
+        }
+    }
+}
+
+/// Status of the node syncing.
+#[derive(Record)]
+pub struct SyncingInfo {
+    /// Ranges of headers that are already synchronised
+    pub stored_headers: Vec<BlockRange>,
+    /// Syncing target. The latest height seen in the network that was successfully verified.
+    pub subjective_head: u64,
+}
+
+impl From<LuminaSyncingInfo> for SyncingInfo {
+    fn from(info: LuminaSyncingInfo) -> Self {
+        Self {
+            stored_headers: info
+                .stored_headers
+                .into_inner()
+                .into_iter()
+                .map(BlockRange::from)
+                .collect(),
+            subjective_head: info.subjective_head,
+        }
+    }
+}
+
+#[derive(Record, Clone, Debug)]
+pub struct PeerId {
+    /// The peer ID stored as base58 string.
+    pub peer_id: String,
+}
+
+impl PeerId {
+    pub fn to_libp2p(&self) -> std::result::Result<Libp2pPeerId, String> {
+        Libp2pPeerId::from_str(&self.peer_id).map_err(|e| format!("Invalid peer ID format: {}", e))
+    }
+
+    pub fn from_libp2p(peer_id: &Libp2pPeerId) -> Self {
+        Self {
+            peer_id: peer_id.to_string(),
+        }
+    }
+}
+
+impl From<Libp2pPeerId> for PeerId {
+    fn from(peer_id: Libp2pPeerId) -> Self {
+        Self {
+            peer_id: peer_id.to_string(),
+        }
+    }
+}
+
+#[derive(Record)]
+pub struct ShareCoordinate {
+    pub row: u16,
+    pub column: u16,
+}
+
+/// Events emitted by the node.
+#[derive(uniffi::Enum)]
+pub enum NodeEvent {
```

Comment:
Can we use the one from `lumina-node` crate?

Reply:
We can, but I haven't yet found a solution to make PeerId from libp2p and Vec<u16,u16> compatible with uniffi without the specific wrappers i wrote in types.rs. Afaik, e.g. tuple types are not directly supported because they cannot be cleanly mapped to equivalent types in different target languages. Of course, I could bring the wrappers into the lumina node if that's wanted :)